R-package/R/lgb.Predictor.R

```diff
@@ -126,6 +127,110 @@ Predictor <- R6::R6Class(
+          stop("Resulting matrix of feature contributions is too large for R to handle.")
+        }
+
+        if (inherits(data, "dsparseVector")) {
+
+          if (length(data) > ncols) {
+            stop(sprintf("Model was fitted to data with %d columns, input data has %.0f columns."
+                         , ncols
+                         , length(data)))
+          }
+          res <- .Call(
+            LGBM_BoosterPredictSparseOutput_R
+            , private$handle
+            , c(0L, as.integer(length(data@x)))
+            , data@i - 1L
+            , data@x
+            , TRUE
+            , 1L
+            , ncols
+            , start_iteration
+            , num_iteration
+            , private$params
+          )
+          out <- new("dsparseVector")
+          out@i <- res$indices + 1L
+          out@x <- res$data
+          out@length <- ncols_out
+          return(out)
+
+        } else if (inherits(data, "dgRMatrix")) {
+
+          if (ncol(data) > ncols) {
+            stop(sprintf("Model was fitted to data with %d columns, input data has %.0f columns."
+                         , ncols
+                         , ncol(data)))
+          }
+          res <- .Call(
+            LGBM_BoosterPredictSparseOutput_R
+            , private$handle
+            , data@p
+            , data@j
+            , data@x
+            , TRUE
+            , nrow(data)
+            , ncols
+            , start_iteration
+            , num_iteration
+            , private$params
+          )
+          out <- new("dgRMatrix")
+          out@p <- res$indptr
+          out@j <- res$indices
+          out@x <- res$data
+          out@Dim <- as.integer(c(nrow(data), ncols_out))
+
+        } else if (inherits(data, "dgCMatrix")) {
+
+          if (ncol(data) != ncols) {
+            stop(sprintf("Model was fitted to data with %d columns, input data has %.0f columns."
+                         , ncols
+                         , ncol(data)))
+          }
+          res <- .Call(
+            LGBM_BoosterPredictSparseOutput_R
+            , private$handle
+            , data@p
+            , data@i
+            , data@x
+            , FALSE
+            , nrow(data)
+            , ncols
+            , start_iteration
+            , num_iteration
+            , private$params
+          )
+          out <- new("dgCMatrix")
+          out@p <- res$indptr
+          out@i <- res$indices
+          out@x <- res$data
+          out@Dim <- as.integer(c(nrow(data), length(res$indptr) - 1L))
+
+        } else {
+
+          stop(sprintf("Predictions on sparse inputs are only allowed for '%s', '%s', '%s' - got: %s"
+                       , "dsparseVector"
+                       , "dgRMatrix"
+                       , "dgCMatrix"
+                       , paste(class(data)
+                       , collapse = ", ")))
+
+        }
+
+        if (NROW(row.names(data))) {
+          out@Dimnames[[1L]] <- row.names(data)
+        }
```

Comment:
Those are already tested in the tests from the PR for row names.

Reply:
They are not. The links in my comment point to tests on `"CsparseMatrix"` objects, but if you click them you won't see tests on the types referenced in this PR: `"dsparseMatrix"`, `"dsparseVector"`, `"dgRMatrix"`, `"dgCMatrix"`.

Is there something I've misunderstood about the relationship between these classes?

Reply:
> They are not. The links in my comment point to tests on `"CsparseMatrix"` objects, but if you click them you won't see tests on the types referenced in this PR: `"dsparseMatrix"`, `"dsparseVector"`, `"dgRMatrix"`, `"dgCMatrix"`.
> 
> Is there something I've misunderstood about the relationship between these classes?

There's a class hierarchy...

Reply:
Please tell us specifically what you mean by "there's a class hierarchy", and why it means that you don't want to add the tests I'm asking you to add.

Reply:
`dgCMatrix` is a subclass of `CsparseMatrix`, which is a subclass of `sparseMatrix`, and so on. Classes like `dsparseMatrix` are abstract.

Reply:
Thanks for that. I'm still struggling to understand how that means that the tests I'm asking for shouldn't be added.

Consider the case added in this PR beginning with `} else if (inherits(data, "dgRMatrix")) {`. It doesn't contain a `return()` statement, so that at the end of the `if - else if` block, the "possibly add row names" code (the line this comment thread is on) will run.

If someone were to add `return(out)` on line 193, that "possibly add row names" code wouldn't be reached. I think it's desirable for a test to fail in that case, to inform us that adding that `return` statement is a breaking change that causes row names to not be set on the predictions.

The tests I linked test a regular dense R matrix and a `CsparseMatrix`. As the example below shows, that means they don't currently cover the cases where the input to `predict()` is a `"dgRMatrix"` or a `"dsparseVector"`.

```r
library(Matrix)

# the first batch of test cases use a regular R dense matrix
X <- matrix(rnorm(100), ncol = 4)
inherits(X, "dgRMatrix")         # FALSE
inherits(X, "dsparseMatrix")  # FALSE
inherits(X, "dsparseVector")  # FALSE
inherits(X, "dgCMatrix")         # FALSE

# the second batch of test cases converts that to a CsparseMatrix
Xcsc <- as(X, "CsparseMatrix")
inherits(Xcsc, "dgRMatrix")         # FALSE
inherits(Xcsc, "dsparseMatrix")  # TRUE
inherits(Xcsc, "dsparseVector")  # FALSE
inherits(Xcsc, "dgCMatrix")         # TRUE
```

Reply:
Ok, if you *really* want to have tests for everything, I've added a test for CSR matrices. A vector representing a single row cannot have row names so I left that out of tests.

Reply:
> if you really want to have tests for everything

Thank you, yes I do. The more of this project's behaviors are reflected in tests, the less likely it is that future changes silently break that behavior. This project is too large for all of these concerns to just be kept in maintainers' heads and enforced through PR comments.